### ISSUE TITLE
Using the aggregator on different downstream task 

### ISSUE THREAD

**Issue by Maram-Helmy** [diffusion-hyperfeatures/diffusion_hyperfeatures#6]
Great work! 

I have a question, what if I want to train the aggregator on different downstream task than correspondence, what should I change? and how it should be trained? I want to use the output feature of the aggregator as a feature representation for a real single image along with its text description. Can you please guide me on this. 

Also, why do you extract the features from the outputs of the residual block before the self- and cross- attention blocks? And in case I'll have a text along the image to be passed to the UNet, then I believe it will be better to extract features after the cross attention features?

Finally, why do you use batch size of two? is it because you are comparing pair of images?

your response is highly appreciated.

Thanks in advance.

**Comment by g-luo** [comment]
Hi!

1. I uploaded a more [generic training script](https://github.com/diffusion-hyperfeatures/diffusion_hyperfeatures/blob/main/train_generic.py). This toy example tries to match the hyperfeatures of a given source and target image, but you can change the dataset and loss according to your needs. You can try it out by running`python3 train_generic.py --config_path configs/train.yaml`.

2. We extract features from the [outputs of the residual block](https://github.com/MichalGeyer/plug-and-play/blob/4e7f703df5f5746011127f058e07ece415564910/run_pnp.py#L112) following [Plug and Play](https://pnp-diffusion.github.io/) (Tumanyan et. al., CVPR 2023). These are also the raw features visualized in the paper, that demonstrate the PCA between objects of the same category have common features (Fig. 3). One could also investigate the features in the cross-attention layers, although we did not in our paper.

3. You can arbitrarily change the batch_size in the config (limited by your GPU VRAM) for extracting features of more images at once. We only assume batch_size=2 for [compute_clip_loss](https://github.com/diffusion-hyperfeatures/diffusion_hyperfeatures/blob/c91869528261698445ee7c580a964d27663a3275/train_hyperfeatures.py#L60) in train_hyperfeatures.py.

**Comment by Maram-Helmy** [comment]
Thanks for your reply!

Can you please provide the code that you used to visualize PCA on the feature maps? So that we can investigate the possibility of using other layers for other tasks.

Your reply is highly appreciated.

**Comment by g-luo** [comment]
Here is our PCA visualization code, which we adapt from Plug and Play.

```
import math
from sklearn.decomposition import PCA

# From https://github.com/MichalGeyer/plug-and-play/blob/403a0fe65bcf34864f1b76cc7dc411b372ec972c/pnp_utils.py#L18
def get_pca(feature_maps_fit_data, feature_maps_transform_data, n_components=3):
    """
    Expects features of shape (B, W*H, C).
    """
    b = feature_maps_transform_data.shape[0]
    h = w = int(math.sqrt(feature_maps_transform_data.shape[1]))
    feature_maps_fit_data = feature_maps_fit_data.reshape(-1, feature_maps_fit_data.shape[-1])
    feature_maps_transform_data = feature_maps_transform_data.reshape(-1, feature_maps_transform_data.shape[-1])
    feature_maps_fit_data = feature_maps_fit_data.cpu().numpy()
    pca = PCA(n_components=n_components)
    pca.fit(feature_maps_fit_data)
    feature_maps_pca = pca.transform(feature_maps_transform_data.cpu().numpy())  # N X 3
    c = n_components
    feature_maps_pca = feature_maps_pca.reshape((b, h, w, c))

    pca_imgs = []
    for i, pca_img in enumerate(feature_maps_pca):
        pca_img = pca_img.reshape(h, w, c)
        pca_img_min = pca_img.min(axis=(0, 1))
        pca_img_max = pca_img.max(axis=(0, 1))
        pca_img = (pca_img - pca_img_min) / (pca_img_max - pca_img_min)
        pca_imgs.append(pca_img)
    return pca_imgs
```

**Comment by Maram-Helmy** [comment]
Thank you so much, for the reply. 

I have another question, what is the dimension of the output of the aggregator? you have mentioned in your paper "we propose a framework for consolidating all intermediate feature maps from the diffusion process, which vary both over scale and time, into **a single per-pixel descriptor** which we dub Diffusion Hyperfeatures" . what is the size/shape of this single per-pixel descriptor? I'm asking this question as I'm trying to match the output of this with text projected in certain dimension. 

Thanks

**Comment by g-luo** [comment]
The dimension of this single per-pixel descriptor (which is the same as the aggregator output) is [batch_size, num_channels, h, w] where num_channels = 384, w = h = 64 in the default config.

The number of channels is a hyperparameter you can play around with, where you could set it differently depending on your text embedding dimension (by changing `projection_dim` in the config). The (h, w) is automatically set to dimension of the input latents to the U-Net.

You should be able to investigate the dimensionality using [get_hyperfeats](https://github.com/diffusion-hyperfeatures/diffusion_hyperfeatures/blob/9b43a65b4c5d87edbcdcfb99a0dd20e9dbfd0014/train_generic.py#L54) in `train_generic.py`.
